### PR TITLE
Fix class name not properly scoped to CLASS

### PR DESCRIPTION
**Pull request by Dominator008**
We need to treat CLASS as a root for a block scope possibly
containing just the "class name" declaration.

See "classScope" in:
http://www.ecma-international.org/ecma-262/6.0/index.html#sec-runtime-semantics-classdefinitionevaluation

Add a few unit tests regarding class names and function names.

Note that our transpilation doesn't support referring to the class name
in a method body yet. Eg:
```js
var Clazz = class Foo {
  bar() {
    return new Foo;
  }
};
```
is currently transpiled to:
```js
/**@constructor @struct */var Clazz = function() {
};
Clazz.prototype.bar = function() {
  return new Foo;
};
```
which does not work at runtime. This will be fixed later.

Part of the fix to #1546.

[Review on Reviewable](https://reviewable.io/reviews/google/closure-compiler/1551)

### CODE REVIEW

File: test/com/google/javascript/jscomp/Es6VariableReferenceCheckTest.java [hunk 1]
```diff
@@ -277,6 +277,10 @@ public void testClass() {
     assertNoWarning("class A { f() { return 1729; } }");
   }
 
+  public void testRedeclareClassName() {
+    assertNoWarning("var Clazz = class Foo {}; var Foo = 3;");
```
Comment (dimvar) [inline test/com/google/javascript/jscomp/Es6VariableReferenceCheckTest.java]: @Dominator008 I would propose to just ban the pattern of using two different names for the same class. What's a good reason to support this?

Comment (Dominator008) [inline test/com/google/javascript/jscomp/Es6VariableReferenceCheckTest.java]: I think the ES6 specs was trying to be consistent with how `function`s look like, i.e. `var func = function f() {}`.

@jleyba Do you have a real-world example showing the usage of this?

Comment (jleyba) [inline test/com/google/javascript/jscomp/Es6VariableReferenceCheckTest.java]: I originally encountered this with a custom Promise implementation. This bug is easy to work around, but the behavior is inconsistent and surprising considering the compiler properly handles scoping on [function names](http://closure-compiler-debugger.appspot.com/#input0%3Dvar%2520Clazz%2520%253D%2520class%2520Foo%2520%257B%257D%253B%250Aclass%2520Foo%2520%257B%257D%250A%250Aclass%2520Bar%2520%257B%257D%250Avar%2520Baz%2520%253D%2520class%2520Bar%2520%257B%257D%250A%250Avar%2520one%2520%253D%2520function%2520two()%2520%257B%257D%253B%250Afunction%2520two()%2520%257B%257D%250A%250Afunction%2520three()%2520%257B%257D%250Avar%2520four%2520%253D%2520function%2520three()%2520%257B%257D%26input1%26conformanceConfig%26externs%26refasterjs-template%26includeDefaultExterns%3D1%26CHECK_SYMBOLS%3D1%26CHECK_TYPES%3D1%26LANG_IN_IS_ES6%3D1%26MISSING_PROPERTIES%3D1%26PRESERVE_TYPE_ANNOTATIONS%3D1%26PRETTY_PRINT%3D1%26TRANSPILE%3D1)

Comment (MatrixFrog) [inline test/com/google/javascript/jscomp/Es6VariableReferenceCheckTest.java]: I agree with @dimvar that we should ban (or at least discourage) this pattern, but I also think we should support it since third party code is already doing this.

Comment (jleyba) [inline test/com/google/javascript/jscomp/Es6VariableReferenceCheckTest.java]: Is it feasible to introduce a more descriptive error message?

Comment (Dominator008) [inline test/com/google/javascript/jscomp/Es6VariableReferenceCheckTest.java]: I think we should support it in the compiler, but discourage it in the style guide.

**Comment by Dominator008** [comment]
I've rebased and resolved the merge conflicts brought by https://github.com/google/closure-compiler/commit/b490b587458f9e26962f32e5e0d5019eaf84b56c.

File: src/com/google/javascript/jscomp/Es6SyntacticScopeCreator.java [hunk 2]
```diff
@@ -100,13 +100,27 @@ private void scanRoot(Node n) {
       }
 
       // Since we create a separate scope for body, stop scanning here
```
Comment (blickly) [inline src/com/google/javascript/jscomp/Es6SyntacticScopeCreator.java]: Looks like this comment is stale

File: src/com/google/javascript/jscomp/Es6SyntacticScopeCreator.java [hunk 3]
```diff
@@ -100,13 +100,27 @@ private void scanRoot(Node n) {
...
+      if (scope.getParent() != null) {
+        inputId = NodeUtil.getInputId(n);
+      }
+
+      final Node classNameNode = n.getFirstChild();
+      // Bleed the class name into the scope, if it hasn't
+      // been declared in the outer scope.
+      if (!classNameNode.isEmpty()) {
+        String className = classNameNode.getString();
+        if (!className.isEmpty() && NodeUtil.isClassExpression(n)) {
```
Comment (blickly) [inline src/com/google/javascript/jscomp/Es6SyntacticScopeCreator.java]: When class has no name, it has an EMPTY node rather than an empty string like function nodes.

File: src/com/google/javascript/jscomp/VariableReferenceCheck.java [hunk 4]
```diff
@@ -227,11 +227,14 @@ private void checkVar(Var v, List<Reference> references) {
                 referenceNode, v);
         boolean letConstShadowsVar = v.getParentNode().isVar()
             && (reference.isLetDeclaration() || reference.isConstDeclaration());
```
Comment (blickly) [inline src/com/google/javascript/jscomp/VariableReferenceCheck.java]: Since the changes to this file look like just refactoring, can we split them into a separate PR?

File: src/com/google/javascript/jscomp/NodeTraversal.java [hunk 5]
```diff
@@ -657,6 +659,41 @@ private void traverseFunction(Node n, Node parent) {
     popScope();
   }
 
+  /** Traverses a class. */
+  private void traverseClass(Node n, Node parent) {
+    Preconditions.checkState(n.getChildCount() == 3, n);
+    Preconditions.checkState(n.isClass());
+
+    final Node className = n.getFirstChild();
+    boolean isClassExpression = (parent != null)
```
Comment (blickly) [inline src/com/google/javascript/jscomp/NodeTraversal.java]: How does this occur? (CLASS node with null parent)

I'd think that would be impossible in a full AST

**Comment by dimvar** [comment]
Sorry, I've dropped the ball on this. Super busy this past week. Reassigning to Ben.

File: src/com/google/javascript/jscomp/Es6SyntacticScopeCreator.java (hunk 2 above)
Comment (Dominator008) [inline src/com/google/javascript/jscomp/Es6SyntacticScopeCreator.java]: It was there since the file was created :) I think it's still right though.

File: src/com/google/javascript/jscomp/NodeTraversal.java (hunk 5 above)
Comment (Dominator008) [inline src/com/google/javascript/jscomp/NodeTraversal.java]: Done. Removed `(parent != null)`. This was copied from `traverseFunction()`, although I'm not sure when `FUNCTION` will have null parent either.

File: src/com/google/javascript/jscomp/Es6SyntacticScopeCreator.java (hunk 3 above)
Comment (Dominator008) [inline src/com/google/javascript/jscomp/Es6SyntacticScopeCreator.java]: Done. Removed the unnecessary check.

File: src/com/google/javascript/jscomp/VariableReferenceCheck.java (hunk 4 above)
Comment (Dominator008) [inline src/com/google/javascript/jscomp/VariableReferenceCheck.java]: Done.

File: src/com/google/javascript/jscomp/NodeTraversal.java [hunk 6]
```diff
@@ -657,6 +659,40 @@ private void traverseFunction(Node n, Node parent) {
...
+    if (!isClassExpression) {
+      // Class declarations are in the scope containing the declaration.
+      traverseBranch(className, n);
+    }
+
+    curNode = n;
+    pushScope(n);
+
+    if (isClassExpression) {
+      // Class expression names are only accessible within the function
```
Comment (blickly) [inline src/com/google/javascript/jscomp/NodeTraversal.java]: function => class

Comment (Dominator008) [inline src/com/google/javascript/jscomp/NodeTraversal.java]: Done.

File: test/com/google/javascript/jscomp/VarCheckTest.java [hunk 7]
```diff
@@ -173,6 +173,7 @@ public void testNamedClass() {
     testSameEs6("class x {}");
     testSameEs6("var x = class x {};");
     testSameEs6("var y = class x {};");
+    testSameEs6("var y = class x { foo() { return new x; } };");
```
Comment (blickly) [inline test/com/google/javascript/jscomp/VarCheckTest.java]: Let's add a test to demonstrate that the class expression name is in scope inside the extends clause.

Comment (Dominator008) [inline test/com/google/javascript/jscomp/VarCheckTest.java]: Done. We can't transpile this code yet though but that's fine.
